File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        count = 0

        if root == None:
            return 0

        def countGoodNodesRec(root, maxVal):
            nonlocal count
            if root == None:
                return
            
            if root.val >= maxVal:
                count += 1
                maxVal = root.val
            
            countGoodNodesRec(root.left, maxVal)
            countGoodNodesRec(root.right, maxVal)

        countGoodNodesRec(root, root.val)
        return count

        

File: Data Structures & Algorithms/delete-leaves-with-a-given-value/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* dfs(TreeNode* curr, int target) {
        if (!curr) {
            return nullptr;
        }

        curr->left = dfs(curr->left, target);
        curr->right = dfs(curr->right, target);

        if (curr->val == target && !curr->left && !curr->right) {
            return nullptr;
        }
        
        return curr;
    }

    TreeNode* removeLeafNodes(TreeNode* root, int target) {
        return dfs(root, target);
    }
};